### PR TITLE
Add new variation modal

### LINKED ISSUE #38884: [ Product Block Editor ] Show add variation modal when `add variation options` is clicked
**Description:**

Blocked by: #39231 
This issue is part of https://github.com/woocommerce/woocommerce/issues/36286 epic.
For testing it would be easier if this [#38882](https://github.com/woocommerce/woocommerce/issues/38882) has been implemented first, otherwise we could use the legacy editor to check.
~~And blocked by https://github.com/woocommerce/woocommerce/issues/38881~~

When a product doesn't have any global or local attributes with the "Used for variations" enabled we should allow the user to add variation options. When a user selects the `add variation options` in the empty state we should show the add variation option modal. This uses the same Modal as the add attributes modal.

We should be able to re-use the add attribute modal.

**Note**
This does **not** include the preview modal on the right and the banner that summarizes what will be created ( see this thread for more context: p1687346601833989/1687346559.131399-slack-C01CMSA1VHS ) 

**Design**

AkVNGImLgSqCObTQ3idVn7-fi-1637_205314

Image

**Acceptance criteria**

- [ ] A add variation modal should show when the `add variation options` button is clicked and add the new options to attributes with "Used for variations" enabled.
- [ ] New attributes should be added to the current product (not saved, as this is not necessary to generate the variations ).
- [ ] Each variation option should be added as a global attribute used for variations. By default, all attributes are created globally. We inform the user about this in the add attributes modal.
- [ ] Any newly created global attribute or global attribute term should be saved ( we can make use of the batch update done as part of https://github.com/woocommerce/woocommerce/issues/39231 )

### PR DESCRIPTION
**Pull request by louwie17**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This adds the new attributes modal to the variations blocks empty state, to allow attributes to be added.

Closes #38884  .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Load this branch and enable the new product editor under **WooCommerce > Settings > Advanced > Features**
2. Install the WooCommerce Beta tester plugin and enable the `product-variation-management` feature under **Tools > WCA Test Helper > Features**
3. Go to **Products > Add New**
4. It should load the new product editor and a **Variations** tab should be present.
5. Go to **Variations** tab and click **Add variation options**
6. A modal should show to add new variation options ( attributes )
7. Add a couple attributes and save the product
8. Refresh the page, the variation options should still be shown on the Variations tab
9. Go to the Organization tab and notice the variations options aren't showing up there.
10. Now make sure you have some global attributes and add the same global attribute to the attributes within the Organization tab.
11. Go back to the Variations tab, the global option should not be present anymore that was selected in Attributes (note: I only did this for global attributes, as local attributes don't have id's, and comparing by name seemed weird )

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: packages/js/product-editor/src/hooks/test/use-product-attributes.test.ts [hunk 1]
```diff
@@ -0,0 +1,387 @@
+/**
```
Comment (louwie17) [inline packages/js/product-editor/src/hooks/test/use-product-attributes.test.ts:1]: Several tests for the `useProductAttributes` filter.

File: packages/js/product-editor/src/hooks/use-product-attributes.ts [hunk 2]
```diff
@@ -78,27 +77,73 @@ export function useProductAttributes( {
...
+			if (
+				attr.id === 0 &&
+				newAttributes.findIndex(
+					( a ) => a.name.toLowerCase() === attr.name.toLowerCase()
+				) !== -1
+			) {
+				return false;
+			}
+			return true;
+		} );
```
Comment (louwie17) [inline packages/js/product-editor/src/hooks/use-product-attributes.ts:115]: Filter the `otherAttributes` so we don't have duplicates and have the server decide which one to keep and which one to remove.

File: packages/js/product-editor/src/hooks/use-product-attributes.ts [hunk 3]
```diff
@@ -78,27 +77,73 @@ export function useProductAttributes( {
...
+			onChange( [
+				...otherAugmentedAttributes,
+				...newAugmentedAttributes,
+			] );
+		} else {
+			onChange( [
+				...newAugmentedAttributes,
+				...otherAugmentedAttributes,
+			] );
+		}
```
Comment (louwie17) [inline packages/js/product-editor/src/hooks/use-product-attributes.ts:137]: Given this hook can be called from a variation field or a `non` variation field, we should keep the order the same.
So in the above I always add variation options at the end.

File: packages/js/product-editor/src/hooks/use-product-attributes.ts [hunk 4]
```diff
@@ -78,27 +77,73 @@ export function useProductAttributes( {
...
 	};
 
 	useEffect( () => {
-		if ( ! getFilteredAttributes().length || attributes.length ) {
```
Comment (louwie17) [inline packages/js/product-editor/src/hooks/use-product-attributes.ts:99]: This logic did not make sense to me, so I removed it. The above prevented changes from being reflected in other fields within the form that were using this hook.
Likewise I removed the `setAttributes` call in the `handleChange` method, and only have it called in this `useEffect`.
This provided more consistency and prevented a infinite loop from happening.

File: packages/js/product-editor/src/hooks/use-product-attributes.ts [hunk 5]
```diff
@@ -78,27 +77,72 @@ export function useProductAttributes( {
...
 			? allAttributes.filter( ( attribute ) => ! attribute.variation )
 			: allAttributes.filter( ( attribute ) => !! attribute.variation );
-		setAttributes( augmentedAttributes );
-		onChange( [ ...otherAttributes, ...augmentedAttributes ] );
+
+		// Remove duplicate global attributes.
+		otherAttributes = otherAttributes.filter( ( attr ) => {
+			if (
+				attr.id > 0 &&
+				newAttributes.findIndex( ( a ) => a.id === attr.id ) !== -1
```
Comment (mdperez86) [inline packages/js/product-editor/src/hooks/use-product-attributes.ts]: You can use `some` function ti check if at least one item exists within an array. 

```suggestion
				newAttributes.some( ( a ) => a.id === attr.id )
```

Comment (louwie17) [reply packages/js/product-editor/src/hooks/use-product-attributes.ts]: Done as part of https://github.com/woocommerce/woocommerce/pull/39522/commits/f8f2b9ef2f5cec258961ddfea82c87595c93d305

**Comment by louwie17** [comment]
@mdperez86 this should be good for a re-review, I addressed your feedback.

**Review by mdperez86 — APPROVED** [review]
LGTM @louwie17 nice job here!!!

**Pull request merged**